_posts/2016-03-28-Mutual-Problems.md

```diff
@@ -0,0 +1,130 @@
+---
+layout: post
+name: mutual-problems
+title: Mutual Problems
+date: 2016-03-28 00:00:00 +00:00
+author: Robert Firek
+image:
+    src: /assets/img/custom/blog/2016-03-28-https.jpg
+    attribution:
+        text: Image by Sean MacEntee (CC BY 2.0)
+        href: https://www.flickr.com/photos/smemon/15944989872/in/photostream/
+tags:
+- Java
+- HTTPS
+- Mutual Authentication
+--- 
+
+The HTTPS protocol is the well-established standard for securing our internet connections. Understanding how this protocol works is not problem and the corresponding [RFC document](https://tools.ietf.org/html/rfc2818) is available since 2000.
+
+Despite HTTPS is used so widely, you can still find a software which doesn't handle this protocol without unnecessary complexity. Unfortunately I'v experienced  problems during the implementation of [the mutual authentication](https://en.wikipedia.org/wiki/Mutual_authentication) in the language which should not surprise me at all. It is **Java**.
+
+## How does HTTPS work?
+
+Before I describe what kind of problems I've got with my implementation, I will describe how the mutual authentication works. The HTTPS protocol use the TLS/SSL protocol to secure the connection. The TLS/SSL protocol defines the authentication handshake which allows to connect any client with the server in the secure way. 
+During [The handshake](https://en.wikipedia.org/wiki/Transport_Layer_Security#TLS_handshake) the following steps are performed:
+
+- A client sends the message which initiates the connection.
+- The server sends its certificate to the client.
+- The client verifies the certificate using the certificate issued by the trusted authority.
+- The server sends the request for the client's certificate.
+- The client sends its certificate to the server.  
+- The server verifies the client's certificate.
+- The server and the client exchange the master secret which is used during the encryption of data.
+- Connection is established.  
+
+Combining my knowledge about the TLS/SSL handshake and my experience from the manual test with `curl` I've assumed that I'm going to need to just three files to implement the client side: **a client's certificate**, **a client's private key** and **a trusted certificate to verify the server's certificate**.
```

Comment:
I assumed that I was going to need only three files to implement the client side

Here you probably want to explain that you only need these three files because you already have the server ready and are only focusing on implementing the client, it's not 100% clear.